From e9539f92b5745ffa1ff47cb9878c0a9d53ed086c Mon Sep 17 00:00:00 2001
From: Chaitanya Kulkarni <kch@nvidia.com>
Date: Sun, 23 Jan 2022 17:05:35 -0800
Subject: [PATCH 14/14] nvme: move io/admin tagset common code to helper

Signed-off-by: Chaitanya Kulkarni <kch@nvidia.com>
---
 drivers/nvme/host/nvme.h | 28 ++++++++++++++++------------
 1 file changed, 16 insertions(+), 12 deletions(-)

diff --git a/drivers/nvme/host/nvme.h b/drivers/nvme/host/nvme.h
index 93e698fbafd1..67141e180466 100644
--- a/drivers/nvme/host/nvme.h
+++ b/drivers/nvme/host/nvme.h
@@ -938,11 +938,11 @@ static inline bool nvme_multi_css(struct nvme_ctrl *ctrl)
 	return (ctrl->ctrl_config & NVME_CC_CSS_MASK) == NVME_CC_CSS_CSI;
 }
 
-static inline void nvme_init_admin_tagset(struct blk_mq_tag_set *set,
-					  void *driver_data, int numa_node,
-					  const struct blk_mq_ops *ops,
-					  unsigned int flags,
-					  unsigned int reserved_tags)
+static inline void __nvme_init_tagset_common(struct blk_mq_tag_set *set,
+					     void *driver_data, int numa_node,
+					     const struct blk_mq_ops *ops,
+					     unsigned int flags,
+					     unsigned int reserved_tags)
 {
 	memset(set, 0, sizeof(*set));
 	set->driver_data = driver_data;
@@ -950,6 +950,15 @@ static inline void nvme_init_admin_tagset(struct blk_mq_tag_set *set,
 	set->ops = ops;
 	set->flags = flags;
 	set->reserved_tags = reserved_tags;
+}
+
+static inline void nvme_init_admin_tagset(struct blk_mq_tag_set *set,
+					  void *driver_data, int numa_node,
+					  const struct blk_mq_ops *ops,
+					  unsigned int flags,
+					  unsigned int rt)
+{
+	__nvme_init_tagset_common(set, driver_data, numa_node, ops, flags, rt);
 	set->queue_depth = NVME_AQ_MQ_TAG_DEPTH;
 	set->nr_hw_queues = 1;
 	set->timeout = NVME_ADMIN_TIMEOUT;
@@ -959,16 +968,11 @@ static inline void nvme_init_io_tagset(struct blk_mq_tag_set *set,
 				       void *driver_data, int numa_node,
 				       const struct blk_mq_ops *ops,
 				       unsigned int flags,
-				       unsigned int reserved_tags,
+				       unsigned int rt,
 				       unsigned int nr_hw_queues,
 				       unsigned int queue_depth)
 {
-	memset(set, 0, sizeof(*set));
-	set->driver_data = driver_data;
-	set->numa_node = numa_node;
-	set->ops = ops;
-	set->flags = flags;
-	set->reserved_tags = reserved_tags;
+	__nvme_init_tagset_common(set, driver_data, numa_node, ops, flags, rt);
 	set->nr_hw_queues = nr_hw_queues;
 	set->queue_depth = queue_depth;
 	set->timeout = NVME_IO_TIMEOUT;
-- 
2.29.0

